src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -2154,22 +2290,37 @@ private MethodSymbol InferMethod(BoundCall node, MethodSymbol method, ImmutableA
             return method;
         }
 
-        private static readonly Func<TypeSymbolWithAnnotations, BoundExpression, BoundExpression> s_makePlaceholderForArgumentFunc =
-            (TypeSymbolWithAnnotations argumentType, BoundExpression argument) =>
+        private ImmutableArray<BoundExpression> GetArgumentsForMethodTypeInference(ImmutableArray<BoundExpression> arguments, ImmutableArray<Result> argumentResults)
+        {
+            // PROTOTYPE(NullableReferenceTypes): MethodTypeInferrer.Infer relies
+            // on the BoundExpressions for tuple element types and method groups.
+            // By using a generic BoundValuePlaceholder, we're losing inference in those cases.
+            // PROTOTYPE(NullableReferenceTypes): Inference should be based on
+            // unconverted arguments. Consider cases such as `default`, lambdas, tuples.
+            return arguments.ZipAsArray(argumentResults, (argument, result) => getArgumentForMethodTypeInference(argument, result));
+
+            BoundExpression getArgumentForMethodTypeInference(BoundExpression argument, Result argumentResult)
             {
+                if (argument.Kind == BoundKind.Lambda)
+                {
+                    // MethodTypeInferrer must infer nullability for lambdas based on the nullability
+                    // from flow analysis rather than the declared nullability. To allow that, we need
+                    // to re-bind lambdas in MethodTypeInferrer.
+                    return GetUnboundLambda((BoundLambda)argument, GetVariableState());
```

Comment:
>GetUnboundLambda((BoundLambda)argument, GetVariableState()); [](start = 27, length = 60)

Would it make sense to do this work just once, rather than once per argument?